### PR TITLE
Make code workflow-compatible, including automated container build

### PR DESCRIPTION
**Pull request by thibaulttabarin**
In gen_metadata.py
change gen_metadata.py with 2 main() option:
  + main_drexel() : original version with minor modification to be used in the container 
  + main_bgnn() : reshape, reformat and simply the output to be used as needed for BGNN_Snakemake.
Modify a few paths (to work with the container)

Add function dataverse_download.py to access model from dataverse read doc
Add Dockerfile for containerization
add github action to build the docker image
Add info to documentation

@DrJPepper @kevinkarnani 
Are you planning to keep developing on this project? Or are you happy to let it go, because if not I am planning to do a huge restructuring of the repo that would look like here https://github.com/thibaulttabarin/BGNN_metadata.

@johnbradley I put you there, just you know what I am up to

### CODE REVIEW

File: Dockerfile
Comment (johnbradley) [inline Dockerfile:69]:
```suggestion
RUN python /pipeline/dataverse_download.py https://datacommons.tdai.osu.edu/ doi:10.5072/FK2/MMX6FY /pipeline/output/
```

File: README.md
Comment (johnbradley) [inline README.md:28]:
```suggestion
python dataverse_download.py https://datacommons.tdai.osu.edu/ doi:10.5072/FK2/MMX6FY output/
```

File: README.md
Comment (johnbradley) [inline README.md:25]:
```suggestion
However this model is not published yet. To download you will need an account on datacommons.tdai.osu and you will need to create an API Token and set the  DATAVERSE_API_TOKEN environment variable.
```

File: dataverse_download.py
Comment (johnbradley) [inline dataverse_download.py:5]: For all of these author comments could we separate out the work into commits? That way you could tell which parts are yours and mine. Especially if there is a bug in the part I wrote :)
If a commit message starting with `Co-authored-by:` it will display the coauthors in the UI. Details:
https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

File: gen_metadata.py
Comment (thibaulttabarin) [reply gen_metadata.py:43]: no, we could remove it but as a lot this code.... We come to the discuss about altering the code too much... If I remove everything that is unnecessary it alters the code too much. I decided to alter the code to the minimum in order to such make it work for the workflow.

File: gen_metadata.py
Comment (thibaulttabarin) [reply gen_metadata.py:175]: same, we could remove it. and what to remove is the question how do we go.... but where do I draw the line... I just modify to a minimum this version of the code. The fork thibaulttabarin/bgnn_metadata is the one with the clean code where I remove as much unnecessary code...

File: dataverse_download.py
Comment (thibaulttabarin) [reply dataverse_download.py:5]: I get the idea and I agree but how to do it exactly... and yes I read the doc.... but does it mean I should reverse the commit and restart at the beginning with co-authorship commit?

File: gen_metadata.py
Comment (thibaulttabarin) [reply gen_metadata.py:45]: how do i add comments in this section as part of the pull request?

File: dataverse_download.py
Comment (thibaulttabarin) [reply dataverse_download.py:5]: if you don't mind I will do next time... I seems too complicated to fix it from here unless you know how to do it

File: gen_metadata.py
Comment (thibaulttabarin) [reply gen_metadata.py:45]: well apparently i have no possibility to act on this...
may be I am wrong but either you suggest a modification and i accept it or I have  fix it in a next commit....
# used for turn off warning message that were not critical

File: gen_metadata.py
Comment (thibaulttabarin) [reply gen_metadata.py:43]: anyway doesn't really matter, either way....

File: README.md
Comment (hlapp) [reply README.md:25]: We should clarify the following things first:
1. Can the model be made public at this point? If yes, we should make it so, and also consider putting it on [Huggingface](https://huggingface.co/imageomics). If not, what are the criteria for making it public.
2. If the model can't be public yet, the description above won't cut it, because people will need not only an account put also be given access.

File: gen_metadata.py
Comment (hlapp) [reply gen_metadata.py]: Commenting out chunks or deleting them outright (if, as here, using version control to do so), is really the same in terms of intrusive code changes.

File: gen_metadata.py
Comment (hlapp) [reply gen_metadata.py]: I notice you fixed it in a subsequent commit.

File: gen_metadata.py
Comment (hlapp) [reply gen_metadata.py:174]: @thibaulttabarin can you explain why you feel this code section needs to be disabled?

**Comment by hlapp** [comment]
I'm putting this PR on hold and @johnbradley will review options for how structure this in a way that minimizes non-generic changes and simplifies the container build. (@thibaulttabarin please do not merge until I approve.)

@DrJPepper and @kevinkarnani I have for now withdrawn the review requests for you. We'll let you know when this is ready again for your review.

**Comment by hlapp** [comment]
I also changed it to draft status to signify it as WIP. Before it goes back to ready for review, @thibaulttabarin we'll also need a proper and descriptive title.

File: README.md
Comment (thibaulttabarin) [reply README.md:25]: 1- make the model public, i say yes but that is not my call, the ball is in Jane Greenberg (@hlapp, you were cc in the email, the response from Jane was unclear). for huggingface, yes that is next on my todo list. 

2- Sure but keep it mind that right now, the point (in priority) is to make this project available and documented for people from our institution BGNN/imageomics. So I assume that people will have (or should have) access to the datacommons/dataverse.

File: gen_metadata.py
Comment (thibaulttabarin) [reply gen_metadata.py:174]: As the code comment say, this is for visualization of some intermediate predictions probably use for development.  
secondly the code create hard coded path for the folders and files.

**Comment by thibaulttabarin** [comment]
@hlapp, @johnbradley, before trying to restructure this in this repository, I would suggested having a look at the restructured (refactored) code I wrote for this project here https://github.com/thibaulttabarin/BGNN_metadata. Especially if this repository is not maintained by drexel anymore (here I refer to the email I send to Jane Greenberg (Hilmar was cc) asking their plan about building on this repository or abandoning it).

F.Y.I. Anyway just saying I put a fair amount of work in the refactoring (this is not completely finished but it is working but I won't finish it unless we are planning to use the code)

File: Dockerfile
Comment (johnbradley) [inline Dockerfile:56]: Why is this Dockerfile using older versions of pytorch/torchvision than the Pipefile already existing in the repo?
https://github.com/hdr-bgnn/drexel_metadata/blob/a8720c3bfefcbd5caf7d948fad4ba90c2adc7f3c/Pipfile#L27-L28

Comment (thibaulttabarin) [reply Dockerfile:56]: In short, the version I used in the Dockerfile is what has worked for me.... I spend quiet sometime testing different version and that what worked. In particular during the development I didn't have access to a GPU and I had to use cpu. Additionally the use of GPU is a bit restrictive (at least it was to me) and since we are using this code as prediction the GPU is not as fundamental.
Bottom line : Everything I did in the changes for this branch is to make my workflow BGNN works.

**Comment by thibaulttabarin** [comment]
I am not sure what I am suppose to do, here. Basically everything I have done, has been reversed.

**Comment by johnbradley** [comment]
The `gen_metadata.py` command line arguments currently are as follows:
```
$ python gen_metadata.py -h
usage: gen_metadata.py [-h] [--outfname OUTFNAME] [--device {cpu,cuda}] [--maskfname MASKFNAME] file_or_directory [limit]

Generate metadata for one or more fish images.

positional arguments:
  file_or_directory     Path to a fish image or a directory of multiple fish images. When one file is passed the JSON
                        metadata is printed to the terminal.
  limit                 Limit the number of images processed from a directory

optional arguments:
  -h, --help            show this help message and exit
  --outfname OUTFNAME   Output filename to use for JSON metadata (disables printing to terminal).
  --device {cpu,cuda}   Override the default device used for the ML model.
  --maskfname MASKFNAME
                        Save a mask image with the provided filename. Only supported when processing a single image file.
```

**Comment by johnbradley** [comment]
I have simplified the changes here and moved the Minnows specific code to the https://github.com/hdr-bgnn/drexel_metadata_formatter repo.

## Summary of Changes

### Adds Docker Container for Metadata Generation
- Adds [Dockerfile](https://github.com/hdr-bgnn/drexel_metadata/blob/release/Dockerfile) that installs requirements from the Pipfile
- Changes the python version in the Pipefile back to the [previous value of "3.8.10"](https://github.com/hdr-bgnn/drexel_metadata/commit/c9667a6c31f075b05ae420ac291f2bf357843e34) to be compatible with the [pytorch requirement in the Pipefile](https://github.com/hdr-bgnn/drexel_metadata/blob/8b64f0bc343efdd839bec9204afa85b59b670187/Pipfile#L27). 
- Downloads the trained model from the [OSU Dataverse](https://datacommons.tdai.osu.edu/) using a multi-stage Docker build.
- Adds GitHub Configuration to automatically [build the docker container](https://github.com/hdr-bgnn/drexel_metadata/blob/release/.github/workflows/deploy-image.yml) when a [GitHub Release is created](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release).

### Adds Command Line Arguments
Adds command line arguments to allow:
- Using CPU instead of GPU
- Specify output metadata filename when processing a single file (instead of printing to the console)
- Specify filename to save fish mask to when processing a single file
- Specify filename to save visualization file when processing a single file

See https://github.com/hdr-bgnn/drexel_metadata/tree/release#running for details.

**Comment by johnbradley** [comment]
I found an additional change needed to be included in a workflow that uses relative file paths for input files.
I'm going to update my Summary of Changes comment above to reflect this addition.

File: Dockerfile
Comment (hlapp) [inline Dockerfile:1]: I'm wondering whether pinning this image at the current version is really a good idea. It means that if we were to update the image as a result of a bug fix or code improvement, we'd have to update the repo here too.

OTOH, we also don't want to risk updating to an image version that is no longer compatible with the behavior expectations codified here.

Perhaps there's some middle ground by promoting `dataverse-access` to 0.1.0, and releasing every 0.1.x image also as version 0.1, which we would pin here. Thoughts?

File: README.md
Comment (hlapp) [inline README.md:110]: Are you sure this shouldn't be `python3` instead of `python`? And the usage command above doesn't prefix with the interpreter, so maybe we don't need to here, either, which would be naturally consistent with the usage command above.

File: README.md
Comment (hlapp) [inline README.md:159]:
```suggestion
J. Pepper, J. Greenberg, Y. Bakiş, X. Wang, H. Bart and D. Breen, "Automatic Metadata Generation for Fish Specimen Image Collections," 2021 ACM/IEEE Joint Conference on Digital Libraries (JCDL), 2021, pp. 31-40, doi: [10.1109/JCDL52503.2021.00015](https://doi.org/10.1109/JCDL52503.2021.00015).
```

File: README.md
Comment (hlapp) [inline README.md:161]:
```suggestion
Kevin Karnani, Joel Pepper, Yasin Bakis et al. Computational Metadata Generation Methods for Biological Specimen Image Collections, 27 April 2022, PREPRINT (Version 1) available at Research Square <https://doi.org/10.21203/rs.3.rs-1506561/v1>
```

File: gen_metadata.py
Comment (hlapp) [inline gen_metadata.py:946]: Shouldn't this say as well _Only supported when processing a single image file_, to be consistent with the other options with this constraint?

Comment (hlapp) [inline gen_metadata.py:946]:
```suggestion
                        help='Output filename to which to print JSON metadata (instead of terminal).')
```

File: gen_metadata.py
Comment (hlapp) [inline gen_metadata.py:942]:
```suggestion
                             'When one file is passed the JSON metadata is printed to the terminal (except see --outfname).')
```

**Review by hlapp — APPROVED** [review]
LGTM. See a few inline comments. All are minor, so I'm approving here proactively. @johnbradley ping here if and when we're ready again for review by Kevin and Joel.

File: Dockerfile
Comment (johnbradley) [reply Dockerfile:1]: Currently the github automation for dataverse-access only tags with `{{major}}.{{minor}}.{{patch}}`. 
Tagging with `{{major}}.{{minor}}` and `{{major}}` should be possible and would be a nice improvement.
Perhaps we should go ahead and create version `1.0.0` since in semver:
> [Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.](https://semver.org/#spec-item-4).

Comment (johnbradley) [reply Dockerfile]: I switched to using the `1` release of `dataverse-access` so rebuilding the container from this repo will receive bug fixes.

**Comment by johnbradley** [comment]
@hlapp The code changes are ready for review by Kevin and Joel.

**Comment by hlapp** [comment]
@kevinkarnani and @DrJPepper I'm re-assigning you for review and sign-off. Please see the [revised changes summary here](#issuecomment-1292602610). I believe the proposed changes are now at the reasonable minimum for integrating the code's capability into a ML workflow.

(As an aside, making the model public is still pending citation / attribution metadata from @janemrc.)

**Review by DrJPepper — APPROVED** [review]
Looks like reasonable cleanup

**Comment by johnbradley** [comment]
@DrJPepper Would you mind creating a new release? That way a docker container with these changes should be built automatically. Thanks.

**Pull request merged**
